config/confighttp/compressor.go

```diff
@@ -23,45 +23,61 @@ type writeCloserReset interface {
 	Reset(w io.Writer)
 }
 
-var (
-	_          writeCloserReset = (*gzip.Writer)(nil)
-	gZipPool                    = &compressor{pool: sync.Pool{New: func() any { return gzip.NewWriter(nil) }}}
-	_          writeCloserReset = (*snappy.Writer)(nil)
-	snappyPool                  = &compressor{pool: sync.Pool{New: func() any { return snappy.NewBufferedWriter(nil) }}}
-	_          writeCloserReset = (*zstd.Encoder)(nil)
-	// Concurrency 1 disables async decoding via goroutines. This is useful to reduce memory usage and isn't a bottleneck for compression using sync.Pool.
-	zStdPool                  = &compressor{pool: sync.Pool{New: func() any { zw, _ := zstd.NewWriter(nil, zstd.WithEncoderConcurrency(1)); return zw }}}
-	_        writeCloserReset = (*zlib.Writer)(nil)
-	zLibPool                  = &compressor{pool: sync.Pool{New: func() any { return zlib.NewWriter(nil) }}}
-	_        writeCloserReset = (*lz4.Writer)(nil)
-	lz4Pool                   = &compressor{pool: sync.Pool{New: func() any {
-		lz := lz4.NewWriter(nil)
-		// Setting concurrency to 1 to disable async decoding by goroutines. This will reduce the overall memory footprint and pool
-		_ = lz.Apply(lz4.ConcurrencyOption(1))
-		return lz
-	}}}
-)
-
 type compressor struct {
 	pool sync.Pool
 }
 
+var (
+	gZipCompressor                    = &compressor{}
+	snappyCompressor                  = &compressor{}
+	zstdCompressor                    = &compressor{}
+	zlibCompressor                    = &compressor{}
+	lz4Compressor                     = &compressor{}
+	_                writeCloserReset = (*gzip.Writer)(nil)
+	_                writeCloserReset = (*snappy.Writer)(nil)
+	_                writeCloserReset = (*zstd.Encoder)(nil)
+	_                writeCloserReset = (*zlib.Writer)(nil)
+	_                writeCloserReset = (*lz4.Writer)(nil)
+)
+
 // writerFactory defines writer field in CompressRoundTripper.
 // The validity of input is already checked when NewCompressRoundTripper was called in confighttp,
-func newCompressor(compressionType configcompression.Type) (*compressor, error) {
-	switch compressionType {
+func newCompressor(compressionType configcompression.TypeWithLevel) (*compressor, error) {
+	switch compressionType.Type {
 	case configcompression.TypeGzip:
-		return gZipPool, nil
+		if gZipCompressor.pool.Get() == nil {
+			gZipCompressor.pool = sync.Pool{New: func() any { w, _ := gzip.NewWriterLevel(nil, int(compressionType.Level)); return w }}
```

Comment:
This seems incorrect. We are taking compressor with _any_ level from the pool. If user picks `zlib/1` first and it goes back to the pool, the second call with any other level like `zlib/2` would still give the same `zlib/1` instance from the pool. We cannot reuse them compressors with different levels.

We might need a map of pools built dynamically here

Reply:
I created a map of compressor pools for compressors that support levels, all the tests pass with this change
https://github.com/open-telemetry/opentelemetry-collector/pull/11084/commits/c7f36ad44861a58491a09f537207a20decb3c186